oak_crypto/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+/// within the same session.
+pub struct ClientCryptoProvider {
+    serialized_recipient_public_key: Vec<u8>,
+}
+
+impl ClientCryptoProvider {
+    /// Creates a new client crypto provider.
+    /// The `serialized_recipient_public_key` must be a NIST P-256 SEC1 encoded point public key.
+    /// <https://secg.org/sec1-v2.pdf>
+    pub fn new(serialized_recipient_public_key: &[u8]) -> Self {
+        Self {
+            serialized_recipient_public_key: serialized_recipient_public_key.to_vec(),
+        }
+    }
+
+    /// Creates an HPKE encryptor by generating an new ephemeral key pair.
+    /// Returns a serialized encapsulated ephemeral public key and a [`ClientRequestEncryptor`].
+    /// The ephemeral public key is a NIST P-256 SEC1 encoded point public key.
+    /// <https://secg.org/sec1-v2.pdf>
+    pub fn create_encryptor(&self) -> anyhow::Result<(Vec<u8>, ClientRequestEncryptor)> {
+        let (serialized_encapsulated_public_key, sender_context, sender_response_context) =
+            setup_base_sender(&self.serialized_recipient_public_key, OAK_HPKE_INFO)
+                .context("couldn't create client request encryptor")?;
+        Ok((
+            serialized_encapsulated_public_key.to_vec(),
+            ClientRequestEncryptor {
+                sender_context,
+                sender_response_context,
+            },
+        ))
+    }
+}
+
+/// Encryptor for client requests that will be sent to the enclave.
+pub struct ClientRequestEncryptor {
+    sender_context: SenderContext,
+    sender_response_context: SenderResponseContext,
+}
+
+impl ClientRequestEncryptor {
+    /// Encrypts `plaintext` and authenticates `associated_data` using AEAD.
+    /// Returns an encrypted request and a corresponding response decryptor.
+    /// <https://datatracker.ietf.org/doc/html/rfc5116>
+    pub fn encrypt(
+        mut self,
+        plaintext: &[u8],
+        associated_data: &[u8],
+    ) -> anyhow::Result<(AeadEncryptedMessage, ClientResponseDecryptor)> {
+        let request = self
+            .sender_context
+            .seal(plaintext, associated_data)
+            .context("couldn't encrypt request")?;
+        let decryptor = ClientResponseDecryptor {
+            sender_context: self.sender_context,
+            sender_response_context: self.sender_response_context,
+        };
+        Ok((request, decryptor))
+    }
+}
+
+/// Decryptor for enclave responses that are received by the client.
+pub struct ClientResponseDecryptor {
+    sender_context: SenderContext,
+    sender_response_context: SenderResponseContext,
+}
+
+impl ClientResponseDecryptor {
+    /// Decrypts `response.ciphertext` and authenticates `response.associated_data` using AEAD.
+    /// Returns a response message plaintext and a request encryptor for encrypting a new request
+    /// within the same session.
+    /// <https://datatracker.ietf.org/doc/html/rfc5116>
+    pub fn decrypt(
+        mut self,
+        response: &AeadEncryptedMessage,
```

Comment:
For symmetry with encrypt it feels like this method should also take separate fields instead of a proto, but up to you